batch_job/src/batch_job_k8s.c

```diff
@@ -0,0 +1,463 @@
+
+static const char *k8s_script_file_name = "_temp_k8s_script.sh";
+
+static const char *k8s_config_tmpl = "\
+{\n\
+    \"apiVersion\": \"v1\",\n\
+    \"kind\": \"Pod\",\n\
+    \"metadata\": {\n\
+        \"labels\": {\n\
+            \"app\": \"%s\"\n\
+        },\n\
+        \"name\": \"%s\"\n\
+    },\n\
+\n\
+    \"spec\": {\n\
+        \"containers\": [{\n\
+            \"name\": \"%s\",\n\
+            \"image\": \"%s\",\n\
+			%s\n\
+            \"imagePullPolicy\": \"IfNotPresent\",\n\
+            \"command\": [\"/bin/bash\", \"-c\"],\n\
+            \"args\": [\"echo \\\"%d,pod_created\\\" > %s.log ; tail -f /dev/null \"]\n\
+        }],\n\
+        \"restartPolicy\": \"Never\"\n\
+    }\n\
+}\n\
+";
+
+static const char *resource_tmpl = "\
+\"resources\": {\n\
+	\"requests\": {\n\
+		\"cpu\": \"%s\",\n\
+		\"memory\": \"%s\"\n\
+	},\n\
+	\"limits\": {\n\
+		\"cpu\": \"%s\",\n\
+		\"memory\": \"%s\"\n\
+	}\n\
+},\n\
+";
+
+typedef struct k8s_job_info{
+	int job_id;
+	char *cmd;
+	char *extra_input_files;
+	char *extra_output_files;
+	int is_running;
+} k8s_job_info;
+
+static k8s_job_info* create_k8s_job_info(int job_id, const char *cmd, 
+		const char *extra_input_files, const char *extra_output_files)
+{
+	k8s_job_info *new_job_info = malloc(sizeof(k8s_job_info));
+	new_job_info->job_id = job_id;
+	new_job_info->cmd = xxstrdup(cmd);
+	new_job_info->extra_input_files = xxstrdup(extra_input_files);
+	new_job_info->extra_output_files = xxstrdup(extra_output_files);
+	new_job_info->is_running = 0;
+
+	return new_job_info;
+}
+
+static batch_job_id_t batch_job_k8s_submit (struct batch_queue *q, const char *cmd, 
+		const char *extra_input_files, const char *extra_output_files, struct jx *envlist, 
+		const struct rmsummary *resources )
+{
+	
+	if(mf_uuid == NULL) {
+		mf_uuid = malloc(sizeof(*mf_uuid));
+		cctools_uuid_create (mf_uuid);
+		// The pod id cannot include upper case
+		string_tolower(mf_uuid->str);
+	}
+	
+	if(k8s_job_info_table == NULL) 
+		k8s_job_info_table = itable_create(0);
+
+	if(k8s_image == NULL) {
+		if((k8s_image = batch_queue_get_option(q, "k8s-image")) == NULL) {
+			fatal("Please specify the container image by using \"--k8s-image\"");
+		}
+	}
+	
+	fflush(NULL);
+	pid_t pid = fork();
+	
+	int job_id = count ++;
+
+	if(pid > 0) {
+
+		debug(D_BATCH, "started job %d: %s", job_id, cmd);
+		struct batch_job_info *info = calloc(1, sizeof(*info));
+		info->submitted = time(0);
+		info->started = time(0);
+		itable_insert(q->job_table, job_id, info);
+		
+		k8s_job_info *curr_job_info = create_k8s_job_info(job_id, cmd,
+				extra_input_files, extra_output_files);
+		
+		itable_insert(k8s_job_info_table, job_id, curr_job_info); 
+		return job_id;
+
+	} else if(pid == 0) {
+
+		if(envlist) {
+			jx_export(envlist);
+		}
+
+		char *pod_id = string_format("%s-%d", mf_uuid->str, job_id);
+		char *k8s_config_fn = string_format("%s.json", pod_id);
+
+		FILE *fd = fopen(k8s_config_fn, "w+");
+		
+		if(resources) {
```

Comment:
A makeflow job could specify just the CPU or just the RAM, right?  So those default needs to be applied independently.

Also, is there a reason to have the default resource size in the k8s driver?  Why not have a default resource selection across all of makeflow?